### ISSUE TITLE
Better control and allow inclusion of the _etag field

### ISSUE THREAD

**Issue by smyrman** [rs/rest-layer#164]
As discussed in #157, there should be a consistent way tocontrol and  include _etag in embedder resources as well as resources in all `GET` views.

**Comment by smyrman** [comment]
@Dragomir-Ivanov , @rs - to continue the discussion here. We have discussed two options previously:

1. Allow `_etag` to ne specified in fields.
2. Allow a separate flag, e.g.  `etag=1`  to include all E-tags.

1 is maybe what gives the most the best feeling; you are planning to update an (embedded) or list item, and you ask for at least the "id" and the "_etag" field. Also it's included as a field, so why not ask for it as a field? However, I have until now assumed that 1 is maybe hard to implement.

If we talk implementation, I have a plausible way forward for 1, but we need to add one more feature to make it work: we need to allow incllusion of hidden fields (and maybe add a new `Secret` field to allow completely hiding). With this done, we add the field it self in the Index during a Resource/Index Bind

**Comment by smyrman** [comment]
Writing on my phone so wikl correct the grammar later...

**Comment by rs** [comment]
I'm all for 1., moving the _tag in the payload and using a hidden/secret/whatever field type to make it not part of the default payload.

**Comment by Dragomir-Ivanov** [comment]
This should govern inclusion the `_etag` into `itemList` case as well.

**Comment by smyrman** [comment]
Some code links.

### Allow hidden fields to be included (could be first PR)
Probably around here


### Allow `_etag` to be specified in as a field (could be second PR)


In `resource.Resource.Bind()`, you can see how `Connection` is added to `r.validator.fallback.Fields`. We will need to do the same for `_etag`, and set the field as `ReadOnly: true, Hidden: true`
https://github.com/rs/rest-layer/blob/3ffe90e45fe2868e3db8dd03c92d0729e9ddeea0/resource/resource.go#L134

We need to do the _same_ in `resource.index.Bind()`:
https://github.com/rs/rest-layer/blob/3ffe90e45fe2868e3db8dd03c92d0729e9ddeea0/resource/index.go#L49

**Comment by Dragomir-Ivanov** [comment]
I think that `Hidden` have to be hidden by default, but can be requested via `fields=` query parameter. This will be inconvenient because in order to get a hidden field, one needs to enumerate all non-hidden fields needed. That is why I wanted this `*` field matcher. 

Currently `Hidden` can also be written when supplied in PATCH. I think that is okay.

The new `Secret` field would not be requested in any way, and will not be available for writes.
Now if we don't specify a field in the resource schema, it will not be available for writes, but will be returned upon GET. `Secret` will prevent this.

Your comments appreciated.

**Comment by smyrman** [comment]
Some minor points.

> This will be inconvenient because in order to get a hidden field, one needs to enumerate all non-hidden fields needed. That is why I wanted this * field matcher.

Agree, but I think it is still good enough. We can have another look at `*` later. Ithink the best approach for `*` is actually to replace it with all non-hidden fields  in the schema _before_ validating fields, and then document the last mention winn rule.

> [Secret field] and will not be available for writes

This should be decided by the exisring `ReadOnly` field. There could be valid use-cases for writing a field without being able to read/expos3 it.

**Comment by smyrman** [comment]
Maybe a better name then `Secret` is `WriteOnly`? It becomes a bit funny (although perfectly valid) if you set both ReadOnly _and_ WriteOnly, but exept for that it might be a better name. Those are allso well-understood concepts in programming...

**Comment by smyrman** [comment]
Another option could be to replace tve ReadOnly field with an enum:

```go
type FieldMode

const (
    ModeReadWrite = iota
    ModeReadOnly
    ModeWriteOnly
    ModeNoAcces
)

**Comment by smyrman** [comment]
We could maybe reuse this for the schema config or vise versa. I actually want to splitt this out to yet another issue. I think the use-case for "secret" / "write-only" fields is probably not to common/strong.

So if we change the meaning of hidden fields for thiss issue without implementing support for what hidden fields used to be, I think that is perfectly OK. @rs , what do you think?

**Comment by Dragomir-Ivanov** [comment]
@smyrman I actually like the last idea about the `FieldMode`. Don't you like to make it just mask:
```
type FieldMode uint
const (
    ModeRead = iota
    ModeWrite
)
```

By default fields can have `ModeRead | ModeWrite`, but then user can overwrite it.
This will make `Hidden` type redundant as well.

@rs Your input appreciated.

**Comment by rs** [comment]
Either way works for me. My preference would go for @smyrman's version as it is more expressive.

**Comment by smyrman** [comment]
> By default fields can have `ModeRead | ModeWrite`

Yeah, I thought about that. However, since it's Go, one extra consideration to have in mind is that all types have default zero values.  Therefore, if you can get away with make the default case a zero value, the code get's signifincantly less complex as no struct initialization method is needed.

If you flip your proposal, and we have `MaskNoWrite = 0x01, ModeNoRead = 0x02`, that would be bitwise equicalent to the iota constants, although perhaps a bit hard to read.

```go
const(
    ModeReadWrite = 0x00
    ModeReadOnly = 0x01 // same as ModeNoWrite
    ModeWriteOnly = 0x02 // same as ModeNoRead
    ModeNoAcces  = 0x03 // same as ModeNoWrite | ModeNoRead
```

> This will make Hidden type redundant as well.

How so? Are we no planning to "redefine" hidden to mean weather or not a field is hidden from the response by default? Won't we still need a field for that?

We could of-course define a bitmask for that as well, but the main goal off the schema should probably be to optimize for the best expressiveness.

**Comment by smyrman** [comment]
Otherwise it was a good proposal @Dragomir-Ivanov.

**Comment by Dragomir-Ivanov** [comment]
@smyrman Yes, that makes sense. I was mistaken, this work will not depreciate `Hidden`.
